### ISSUE TITLE
UI Bootstrap accessibility

### ISSUE THREAD

**Issue by jharris-westpac** [WestpacCXTeam/Angular-GUI#2]
I want to make this thread about accessibility and trying to find a way to improve our current setup. 

In our current efforts to port the GUI 2 code into our WPF AngularJS codebase, we have had to re-use some of the existing UI bootstrap components. UI bootstrap based entirely on Bootstrap 3 markup, has issues with accessibility. We therefore need to define clearly what those issues are, how to create and run test cases for each issue and how to build new components in an accessible way. 

The GEL 2 base has been certified as accessible AA?, so this thread is for components which deviate from the GEL 2 HTML base.

**Comment by jharris-westpac** [comment]
@dominikwilkowski Do you do any of the accessibility testing yourself? What is the hardware that you are using in order to do these tests so we can better improve? Saying an entire app or framework is not accessible is not really helpful as we need specifics. We need a way forward which is why we the developers and testers would need the tools at hand in development to make our apps accessible. We will often write custom components and need to make these accessible and testable too. If we cant create an isolated reproducible accessibility test case for an HTML component, we have no way forward. 

I personally have seen way to many highly opinionated articles on this subject which is why I am looking for a concrete reference which we can articulate, plan and implement.

@tonytwong Adding you in as the SME for WPF.

**Comment by dominikwilkowski** [comment]
:+1:

GUI 2.0 has gone through a formal accessibility report and passed. We do some testing ourself via NVDA, JAWS and Voiceover. No special hardware needed. There are great resources out there to help you with what is accessible and what is not but your best bet is always to write semantic code as that's what screen readers are based on. We also don't just check for screenreader accessibility but also for keyboard users, users with visually impairment and cognitive impairments.

The standard procedure is to engage Vision Australia to do a complete report on what you have so far and give you suggestions. We worked with Russ Weakley and Roger Hudson separately who are both amazing! Roger has been engaged by tech several times before.

I am not aware of any way you can automate accessibility testing entirely. Let me know if you find a way.

Let me know if I can help any other way.

**Comment by jharris-westpac** [comment]
Thanks @dominikwilkowski sounds good. We will choose a desktop screen reader and mobile (Voiceover) and start running with those and record results. 

Did Vision Australia provide a detailed report on areas they checked and how to improve?  

We can make sure our code is semantic and validated at a minimum as any good dev should :+1:   

I may recommend to the business that we get at least 1 dedicated accessibility tester on each floor if this is a requirement per Westpac standards and we can ask Vision Aus if they provide training. If we cant automate it by hardware and its a matter of interpretation then we need to get a dev or tester the right training. I don't believe its enough to have a product tested at the end of development, it should be at regular intervals throughout the dev lifecycle so we need a human or a  machine actively involved in the project.

**Comment by dominikwilkowski** [comment]
I can't agree with you more :)
SPAs all have some additional requirements around accessibility and I saw some of the reports for BBO(an angular app).

The report usually tells you what fails and recommends something but that's different with every tester.
Things they test:
- screenreaders
  - is everything announced
  - are there shortcuts to jump to the important bits?
  - are the announcements sensible?
  - are changes announced
- font-size increase
- zoom
- is focused move to the appropriate elements(trap focus etc)
- arrow keys and tab keys functionality where expected
- responsiveness
- overall usability
- color contrasts
- page without CSS
- page without JS(in your case a simple warning is enough)
- focusability ie keyboard users

and more.

**Comment by oksushi** [comment]
Just throwing [Tenon](http://tenon.io) in the mix — it’s not a replacement for an accessibility review by a knowledgeable human, but it does automate a lot of it. They have an API you can add into your build process to run tests, too.

**Comment by tonytwong** [comment]
@jharris-westpac suggest NVDA (http://www.nvaccess.org/) as your desktop screen reader (it is free) as a sanity check for accessibility to begin with, and where access exists to JAWS screen reader licence we then take further testing from there. Beyond that you will find on occasion project will engage Vision Australia (accessibility consultants) to further test the application and surface any other accessibility concerns. 

Not specific to angular ui-bootstrap, but paypal have documented some tweaks to enhance boostrap markup to get some further accessibility into bootstrap modules which you can possibly implement. To be honest the ui-bootstrap team may have already implement some of these into the latest build. but this is good for reference of exisitng legacy modules in your project. https://github.com/paypal/bootstrap-accessibility-plugin#details

**Comment by jharris-westpac** [comment]
Thanks @tonytwong will take a look. On Thursday the guys here are organizing for myself and our tester Geetha to go to barangaroo and sit with a blind accessibility tester and go through our application. I think it will be great to see how they are using it and where we can practically improve.

**Comment by dominikwilkowski** [comment]
Hey guys,

How is it going? Is this moving forward?

On another note: For our efforts to move the GUI into submodules I have to rename this repo. Shouldn't effect you at all as GitHub keeps the old SSH links. Just an FYI

**Comment by jharris-westpac** [comment]
Hi, the BankWow Switching app managed to get approval from the MAD accessibility testers and we went into production this weekend. It is usable by screen readers but room for improvement. 

Key learning points were:
- Modals were very difficult to get right across all the screen readers, just adding role, and an auto focus did not always do it and I still find it temperamental, I found faults in GEL 1, GEL 2, PayPal accessible modals, basically everything, so i can only think its buggy readers. The primary issue being it can intermittently start reading the content behind the modal. But not often, and tends to be modals which automatically open. 
- Bootstrap components are a little difficult to modify, much of the content is hard coded and the JS behind them is fairly complex. 
- NVDA and me do not get along, perhaps its just a very complex thing to use, but beyond tabbing around, I found its lacking basic modes like reading the next paragraph in which iOS does fine. I would prefer to throw NVDA away and get a licence for JAWS. I dont mind the iOS VoiceOver as at least it was consistent and straight forward to use. 
  - Our accessibility testers in the MAD team did not know how to use the Screen readers correctly and were raising numerous defects which were not defects at all and intentional read outs. We need anyone who is testing this stuff to know how to use the tools properly and have correct assertions as per the standards guidelines. W3C has clear guidelines on how readers should read pages. 
  - AngularJS apps work pretty well surprisingly, I am told that Moz Firefox is the best browser when using a screen reader on Desktop. Angular page changes need to have the headings re-focused on each route change so the reader knows the page has changed and to start reading from the top again. 
  - Our SVG checkbox and radios should have the focus outline set so you know which object has the focus. Other inputs have the default browser outline which works fine. 
- The angular-aria module is partially useful, but breaks tabindex on checkbox groups, you can override and turn off that feature. 

Next steps is to see if we can get a licence for JAWS, as I suspect that is much better to use than NVDA. We need to talk to product owner and some people higher up to see if we can get a genuine accessibility expert on the floor.

**Comment by dominikwilkowski** [comment]
Sounds great @jharris-westpac 
Let me know if I can do anything to support the accessible efforts.
